client/trino-jdbc/src/test/java/io/trino/jdbc/TestJdbcPreparedStatement.java

```diff
@@ -1051,128 +1074,225 @@ public void testConvertTimestamp()
+        assertBind((ps, i) -> ps.setObject(i, String.class), useLegacyPreparedStatements).isInvalid("Unsupported object type: java.lang.Class");
+        assertBind((ps, i) -> ps.setObject(i, String.class, Types.BIGINT), useLegacyPreparedStatements)
+                .isInvalid("Cannot convert instance of java.lang.Class to SQL type " + Types.BIGINT);
+        assertBind((ps, i) -> ps.setObject(i, "abc", Types.SMALLINT), useLegacyPreparedStatements)
+                .isInvalid("Cannot convert instance of java.lang.String to SQL type " + Types.SMALLINT);
+    }
+
+    @Test
+    public void testLegacyPreparedStatementsTrue()
+            throws Exception
+    {
+        testLegacyPreparedStatementsSetting(true,
+                "EXECUTE statement1 USING INTEGER '42', 'value1''s'");
     }
 
-    private BindAssertion assertBind(Binder binder)
+    @Test
+    public void testLegacyPreparedStatementsFalse()
+            throws Exception
     {
-        return new BindAssertion(this::createConnection, binder);
+        testLegacyPreparedStatementsSetting(false,
+                "EXECUTE IMMEDIATE 'SELECT * FROM blackhole.blackhole.test_table WHERE x = ? AND y = ? AND y <> ''Test''' USING INTEGER '42', 'value1''s'");
     }
 
-    private Connection createConnection()
+    private BindAssertion assertBind(Binder binder, boolean useLegacyPreparedStatements)
+    {
+        return new BindAssertion(() -> this.createConnection(useLegacyPreparedStatements), binder);
+    }
+
+    private Connection createConnection(boolean useLegacyPreparedStatements)
             throws SQLException
     {
         String url = format("jdbc:trino://%s", server.getAddress());
-        return DriverManager.getConnection(url, "test", null);
+        Properties properties = new Properties();
+        properties.setProperty("user", "test");
+        properties.setProperty("legacyPreparedStatements", String.valueOf(useLegacyPreparedStatements));
+        return DriverManager.getConnection(url, properties);
     }
 
-    private Connection createConnection(String catalog, String schema)
+    private Connection createConnection(String catalog, String schema, boolean useLegacyPreparedStatements)
             throws SQLException
     {
         String url = format("jdbc:trino://%s/%s/%s", server.getAddress(), catalog, schema);
-        return DriverManager.getConnection(url, "test", null);
+        Properties properties = new Properties();
+        properties.setProperty("user", "test");
+        properties.setProperty("legacyPreparedStatements", String.valueOf(useLegacyPreparedStatements));
+        return DriverManager.getConnection(url, properties);
+    }
+
+    private void testLegacyPreparedStatementsSetting(boolean legacyPreparedStatements, String expectedSql)
+            throws Exception
+    {
+        Method getExecuteSql = TrinoPreparedStatement.class.getDeclaredMethod("getExecuteSql", String.class, List.class);
```

Comment:
There is a less intrusive way to check that we're issuing the expected statements - query `system.runtime.queries`.

Reply:
Test assertions updated accordingly